src/store/ChartStore.js

```diff
@@ -272,6 +281,62 @@ class ChartStore {
         this.feed.onComparisonDataUpdate(this.updateComparisons);
     }
 
+    /**
+     * Get tradeTimes if not loaded yet
+     * OR update the active symbols by comapring open time
+     */
+    @action.bound updateMarketClosedStatus() {
+        const nowUtc = (new Date()).getTime();
+        const toEpochGMT = (hour, crossDay) => {
+            const currentDate = new Date();
+            currentDate.setDate(currentDate.getDate() + (crossDay || 0));
+            const dateStr = currentDate.toISOString().substring(0, 11);
+            return new Date(`${dateStr}${hour}Z`).getTime();
+        };
+
+
+        if (this.tradeTimes.length && this.tradeTimesCapture.getDate() === (new Date()).getDate()) {
+            this.tradeTimes.forEach((market) => {
+                market.submarkets.forEach((submarket) => {
+                    submarket.symbols.forEach((symbol) => {
+                        const foundSymbol = this.activeSymbols.find(item => item.symbol === symbol.symbol);
+                        if (foundSymbol) {
+                            let isOpen = false;
+                            for (let i = 0; i <= symbol.times.open.length; i++) {
+                                const { open, close } = symbol.times;
+                                if (open.length && close.length) {
+                                    const openTime = toEpochGMT(open[i]);
+                                    const closeTime = toEpochGMT(close[i]);
+
+                                    // If open time is cross day, then should check time till tomorrow
+                                    // and yesterday till now
+                                    if (
+                                        (openTime > closeTime && nowUtc >= toEpochGMT(open[i], -1) && nowUtc <= closeTime) ||
+                                            (openTime > closeTime && nowUtc >= toEpochGMT(open[i]) && nowUtc <= toEpochGMT(close[i], 1)) ||
+                                            (nowUtc >= openTime && nowUtc <= closeTime)
+                                    ) {
+                                        isOpen = true;
+                                        break;
+                                    }
+                                }
+                            }
+                            foundSymbol.exchange_is_open = isOpen;
+                        }
+                    });
+                });
+            });
+
+            this.categorizedSymbols = this.categorizeActiveSymbols();
+        } else {
+            this.api.getTradingTimes().then((data) => {
+                this.tradeTimesCapture = new Date();
+                this.tradeTimes = data.trading_times.markets;
+                this.updateMarketClosedStatus();
```

Comment:
@shayanbinary this function is complicated for not good reason, you don't really need `this.tradeTimes` & that check the if clause (`if( ... && this.tradeTimesCapture.getDate() === (new Date()).getDate())`) is unnecessary.

```js
setInterval(() => {
    const trading_times_promise = this.api.getTradingTimes();
    trading_times_promise.then(this.updateMarketClosedStatus)
                                           .catch(data => console.warn(data.error));
}

// and later here
@action.bound updateMarketClosedStatus(tradingTimes) {
    // simplified implementation
}
```

Reply:
@aminroosta  well main purpose of this piece of code to get the trading time each `day`, and update the list each `10 minutes`. we just try to prevent unnecessary API call

Reply:
@shayanbinary True, but you are calling `this.api.getTRadingTimes()` every 10 min.

Reply:
@shayanbinary what do you mean by "prevent unnecessary API calls"  here? where is the unnecessary call?

Reply:
@aminroosta at first I call API and fill the `this.tradeTimes` then in the time that user is date not change with the *last API Call date* (`this.tradeTimesCapture.getDate() === (new Date()).getDate()`) I just update the trade closed status, if date change, the make new api call

Reply:
@shayanbinary that optimization should go to `api.getTradingTimes()` (in `BinaryAPI.js`).

Reply:
@aminroosta  so you mean I do the same optimization as I did here on  the `BinaryAPI.js`. I mean store data there and return to user if user call the api within `1 day` or if it's within 1 day just don't return anything

Reply:
@shayanbinary for now just call `api.getTradingTimes()`, don't worry about extra requests.

Later we will optimize `getTradingTimes()` to cache promises (not the data).
Something like

```js
   getTradingTimes() {
       const yyyymmdd = new Date().toISOString().substring(0, 10);
        return this.requestAPICached({
            trading_times: yyyymmdd,
        });
    }

    cached = {};
    requestAPICached(req) {
        const key = JSON.stringify(req);
        if(cached[key]) return cached[key];
        const promise = requestAPI(req);
        cached[key] = promise;
        promise.catch(() => { delete cached[key]; });
        return promise;
    }
```

Reply:
@brucebinary while writing the above comment i realized the following BinaryAPI.js implementation is incorrect.

```js
    async getActiveSymbols() {
        return this.requestAPI({
            active_symbols: 'brief',
            product_type: 'basic',
        });
    }

    async getTradingTimes() {
        return this.requestAPI({
            trading_times: 'today',
        });
    }
```

All those `async` keywords are useless (& probably a syntax error according to es7 rules), because an async function by definition should `await` for something (a promise).

probably the compiler is generous heree, it just ignores such wrong use cases of `async` keyword.